anxcloud/data_source_vlan.go

```diff
@@ -0,0 +1,118 @@
+			"vm_provisioning":      {Type: schema.TypeBool, Computed: true, Description: "True if VM provisioning shall be enabled. Defaults to false."},
+			"description_customer": {Type: schema.TypeString, Computed: true, Description: "Additional customer description."},
+			"description_internal": {Type: schema.TypeString, Computed: true, Description: "Internal description."},
+			"role_text":            {Type: schema.TypeString, Computed: true, Description: "Role of the VLAN."},
+			"status":               {Type: schema.TypeString, Computed: true, Description: "VLAN status."},
+			"locations":            schemaLocations(),
+		},
+	}
+}
+
+func dataSourceVLANRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c := m.(providerContext).legacyClient
+	v := vlan.NewAPI(c)
+
+	var (
+		id, _    = d.Get("id").(string)
+		name, _  = d.Get("name").(string)
+		resultID = ""
+	)
+
+	// The schema ensures that either ID or name have to be set. We first check
+	// whether the identifier has been set and if so, return the VLAN for that.
+	// Otherwise, we fall back to search by name.
+	if id != "" {
+		byID, diags := findVLANByID(ctx, v, id)
+		if diags.HasError() {
+			return diags
+		}
+		resultID = byID
+	} else if name != "" {
+		byName, diags := findVLANByName(ctx, v, name)
+		if diags.HasError() {
+			return diags
+		}
+		resultID = byName
+	} else {
+		return diag.Errorf("Either provide a non-empty %q or %q to query a VLAN.", "id", "name")
+	}
+
+	// Once we determined our identifier, we set it and reuse our code for fetching the detailed information.
+	d.SetId(resultID)
+	return resourceVLANRead(ctx, d, m)
+}
+
+// findVLANByName searches for a given VLAN by the provided name.
+func findVLANByName(ctx context.Context, v vlan.API, name string) (string, diag.Diagnostics) {
+	var foundID string
+
+	vlans, err := listAllPages(func(page int) ([]vlan.Summary, error) {
+		return v.List(ctx, page, 100, url.QueryEscape(name))
+	})
+	if err != nil {
+		return "", diag.FromErr(fmt.Errorf("querying VLAN with name %q from engine: %w", name, err))
+	}
+
+	// Because searches like "VLAN123" would also match "VLAN1234", we iterate over
+	// the results and search for an exact match.
+	for _, v := range vlans {
+		if v.Name == name {
+
+			// Although VLAN identifiers should be unique, we should not rely on that. In
+			// case there's an ambiguity, better report an error instead of accidentally
+			// breaking infrastructure.
+			if foundID != "" {
+				return "", diag.Errorf("Name ambiguity detected when searching for VLAN with name %q. You should reference the VLAN using one of its identifiers (%s) instead of relying on the name.",
+					name,
+					strings.Join([]string{foundID, v.Identifier}, ", "))
+			}
+
+			foundID = v.Identifier
+		}
+	}
+
+	// If the ID is still empty, that means that we haven't found a suitable VLAN.
+	if foundID == "" {
+		return "", diag.Errorf("No VLAN found with the name %q.", name)
+	}
+
+	return "", nil
```

Comment:
This could reduce the number of returns
```suggestion
	var diags diag.Diagnostics

	// If the ID is still empty, that means that we haven't found a suitable VLAN.
	if foundID == "" {
		diags = append(diags, diag.Errorf("No VLAN found with the name %q.", name)...)
	}

	return foundID, diags
```